src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs

```diff
@@ -205,7 +204,17 @@ public static void SetStatusTrailers(HttpResponse response, Status status)
         {
             // Use SetTrailer here because we want to overwrite any that was set earlier
             SetTrailer(response, GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
-            SetTrailer(response, GrpcProtocolConstants.MessageTrailer, status.Detail);
+
+            var escapedDetail = status.Detail;
+            if (!string.IsNullOrEmpty(status.Detail))
+            {
+                // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+                // The value portion of Status-Message is conceptually a Unicode string description of the error,
+                // physically encoded as UTF-8 followed by percent-encoding.
+                escapedDetail = Uri.EscapeDataString(status.Detail);
```

Comment:
qq about performance: if original the string doesn't contain any characters that would need to be percent-encoded,  is the same original string object returned?

Reply:
Yup: https://github.com/dotnet/corefx/blob/a6f76f4f620cbe74821c6445af3f13e048361658/src/System.Private.Uri/src/System/UriExt.cs#L586-L587. The EscapeString method returns null if nothing is escaped.